### ISSUE TITLE
Bounty for a fully native Bisq app for Android

### ISSUE THREAD

**Issue by wiz** [bisq-network/proposals#139]
> _This is a Bisq Network proposal. Please familiarize yourself with the [submission and review process](https://docs.bisq.network/proposals.html)._

## Background

Over the past month I’ve been working with @bodymindarts on [The Risq Project](https://github.com/bodymindarts/risq), a feasibility study on creating an alternative implementation of Bisq in rust language. Since rust compiles to native code, we could achieve much better performance than Java, which opens up many new potential use cases for Bisq. I'm very proud that Risq is quickly making progress and we're already using Risq in production for the Bisq Markets API service and my infrastructure alerts system to monitor Bisq seednodes.

But the coolest breakthrough in our feasibility study was when I had the crazy idea to ask @bodymindarts to try using Android's Native Development Kit to package Risq and Tor into a Proof of Concept native Android app, and [he successfully got it working](https://github.com/bodymindarts/risq-android) in only a few days! After that, we started using the PoC Android app to study how well we can directly connect to the Bisq P2P network from an Android device, and observe what the CPU, RAM, battery, and network usage would look like.

## The Proof of Concept

<img width="450" alt="Screenshot 2019-11-10 at 15 25 46" src="https://user-images.githubusercontent.com/232186/68560379-2f101a00-0484-11ea-875f-c20563653396.png">

After some testing, we realized it's totally possible to create a native Bisq for Android app using Risq and Tor, without the need for a desktop computer or even a Raspberry Pi at home. You could even be at a bar or Bitcoin meetup and have someone to install Bisq on their phone and instantly start trading with them on Bisq!

## The Dream

I asked @pedromvpg to design some UI mockups of what Bisq might look like on Android, and he quickly sent over the following beautiful work

<img width="210" src="https://user-images.githubusercontent.com/232186/68560395-3800eb80-0484-11ea-808b-dfdc363a3493.png"> <img width="210" src="https://user-images.githubusercontent.com/232186/68560397-3800eb80-0484-11ea-9db8-82ccc3d97094.png"> <img width="210" src="https://user-images.githubusercontent.com/232186/68560398-38998200-0484-11ea-9d10-b312dba5fec5.png">

## The Estimate

To implement basic trading functionality (specs below), our rough estimate is that it will take 3 people working full-time about 6 months to implement all of the low-level functionality required, together with a nice UI / UX that average users would be able to use. We could then publish the Bisq for Android app on the Google Play store, and also allow direct APK installations from the Bisq website for censorship resistance.

## The Proposal

Since the Bisq DAO's compensation system only pays for contributions after they are completed and delivered, it's not well suited for long-term projects like a full-blown Android app which would take 6+ months to complete. For large projects like this, it's necessary to have a team leader fund the development of the project out of their own pocket, until the project is completed and request compensation to reimburse the expenses. If this proposal is accepted, I will fund our development fully and only request compensation after the Android app is completed to the satisfaction of the Bisq DAO.

## The Specifications

* Single Android APK bundling TOR + Risq + Native android UI
* Bitcoin support for the trade wallet via BIP 158 / 157 (aka Neutrino protocol)
* Minimal trade functionality using external Bitcoin wallet app installed on Android device
* Minimal trade dispute resolution
* No BSQ wallet or DAO / Governance (will be added in a later update and not part of this proposal)

## The Bounty

Upon acceptance by the Bisq DAO, I will begin funding the project and take all the risk myself, and only after completion I will request the bounty be paid out to reimburse the cost of development for the project. The bounty amount will be $180K USD to compensate 3 people working full-time for 6 months at an average rate of $10K/month/person.

If the project is not completed or does not work for some reason, the Bisq DAO will not compensate us, so this is a no-risk proposal for Bisq.

## Try it out yourself

Please [download and install the Proof of Concept Bisq for Android app](https://github.com/bodymindarts/risq-android/releases) to see how far we've gotten. Give it a minute to start up (no loading screen yet!) and imagine how cool a fully functional Bisq app for Android would be to easily onboard new users :grin:

**Comment by niyid** [comment]
Good work @wiz,  @bodymindarts

Maybe you can create a Slack channel (if not done yet) on this as this is a project of highest priority and I will be glad to join. I have long been looking at the fastest route to get Bisq on Android and other Mobile devices.

**Comment by bodymindarts** [comment]
@niyid we talk on the #risq channel on bisq org in keybase. Slack is too vulnerable to impersonation. You're welcome to join us there.

**Comment by niyid** [comment]
> @niyid we talk on the #risq channel on bisq org in keybase. Slack is too vulnerable to impersonation. You're welcome to join us there.

Great. Thanks.

**Comment by chimp1984** [comment]
@wiz 
I highly appreciate the initiative to push the project to get an Android app developed and your commitment to pre-fund such a project!

My only concern with that particular proposal is the dependency on the Rust (risq project) implementation. I expressed my concerns regarding risq here: https://github.com/bisq-network/proposals/issues/125#issuecomment-553233280.

I would suggest an alternative approach:

- Build on top of the upcoming Bisq API (either GRPC or HTTP API) 
- Use the same architecture of a headless server app and a UI client with flexible setup so the mobile user can run the full Bisq app on their phone if the headless Bisq app is running there or they can run the headless app on any server/cloud service or they connect to their Desktop app which is configured to run in server mode. 

For reducing resource requirements the user can deactivate the DAO by an option key. Ram usage is roughly 40% lower as a quick test has shown. 
Improving performance is on our top prio list anyway, so expect lower resource requirements soon. BitcoinJ is currently our main resource consumer but as soon we have enough dev resources that will likely be replaced by Electrum which will result in much lower resource requirements and faster startup time.

I think using the existing headless Bisq app as basis comes with many benefits:
- You can use it soon and dont need to wait 6-12 months for a ”risqy" re-implementation
- No extra costs for building on top of Bisq by using the API instead of reimplementing Bisq
- It is the same code base as the desktop App so maintenance is much easier  
- The DAO can be enabled or disabled by a flag
- No risk of consensus or compatibility issues
- Any improvements on the Java implementation will flow directly into the headless app as well (same code base)
- New featues will require only updates in the UI not in the business domain code.

I am not much familiar with Android development but if the packaging of the Bisq headless app into Android is a challenge maybe we should start with a bounty task to see if that is feasible. I expect and hope that this should not be a major problem.

Alternatively to pre-fund the full project you could consider to provide a roadmap with clearly defined acceptance criteria which once they are met get compensated. We should discuss in the context of the incubator idea how we can deal with such situations.

Please note, that I will reject the current proposal in the DAO voting as I interpret it as being based on the risq project which I do not support. But I highly appreciate the initiative to get a mobile app out using the Bisq API approach and would fully support work on that.

**Comment by bodymindarts** [comment]
>  My only concern with that particular proposal is the dependency on the Rust (risq project) implementation. I expressed my concerns regarding risq here: #125 (comment).

@chimp1984 many of your concerns you bring up in your comment on that proposal are not actually relevant for V2. They are however relevant for this proposal as this somewhat implies at least a partial re-implementation of V1. I think it would be very helpful if we could keep those discussions separate.

**Comment by bodymindarts** [comment]
> I am not much familiar with Android development but if the packaging of the Bisq headless app into Android is a challenge maybe we should start with a bounty task to see if that is feasible. I expect and hope that this should not be a major problem.

I would consider running Bisq standalone on Android unfeasible until proven otherwise. The resource requirements of risq are almost an order of magnitude less than Bisq in every dimension. In some cases the gain is >200% (eg RAM usage is constant @ ~100mb of RAM vs 1.7GB after start up which grows to > 2GB after 24hours on my 2015 MBP running latest OSX).

**Comment by chimp1984** [comment]
@bodymindarts Bisq can run with about 200 MB of RAM. I had about 30-40 instances running of my machine for testing purposes and memory went down to about 200 MB per instance. Java is greedy and use what the OS has available on RAM... Your risq app cannot be compared with Bisq yet as it has implemented only a tiny part of it yet.

**Comment by trigger67** [comment]
@wiz I discovered Bisq very recently, but I like your idea and think that a mobile app would be amazing. There is just one thing that I dislike: it would require a lot of work, just for Android platform...
**Have you considered using web technologies instead of developing a native android app ?**

I am no expert, but I think developing a Progressive Web App would be the best option (if possible). It would work just like a native app on any OS. All that with one single code base... (I will elaborate later if this is considered.)

Is it possible ?
I know that rust can be compiled to web assembly, so you can use Risq in a PWA.
For connecting to Tor network and protecting privacy I have no clue... But could try to find the information, or help to develop a PoC.

**Comment by niyid** [comment]
@chimp1984 @bodymindarts 

I will find a memory performance/footprint report comparing both Java and Rust implementations interesting. Maybe creating a stripped down version of Bisq comparable to the Risq version will help.

**Comment by yoshimo** [comment]
i partially agree with @chimp1984 , we should have a set of intermediate goals that can be compensated but i think moving away from java to a secure language with rust is still a goal that should be reached for all platforms. 

Having to front the money for people is quite a burden for a person and i think if we split it across multiple people in the community we might be able to reach the goal faster depending on how many people are intrested in this.

**Comment by bodymindarts** [comment]
> @bodymindarts Bisq can run with about 200 MB of RAM. I had about 30-40 instances running of my machine for testing purposes and memory went down to about 200 MB per instance. Java is greedy and use what the OS has available on RAM... 

Thats very interesting. If that is the case its probably worth trying! My intention with risq is an investigation into implementing V2 anyhow. For V1 on android running headless bisq is probably preferable - if it is possible.

**Comment by wiz** [comment]
@chimp1984 Since we weren't aware of your work on the Bisq java app's headless GRPC API implementation when we started the Android app PoC using risq, we didn't consider this option. You have a lot of very good points, and I discussed your suggestions with @bodymindarts and we agreed to accept your suggestions :grin:

I propose that we amend this DAO proposal as follows:
- we put the risq project on hold
- we develop the Android app using Bisq java headless GRPC API (technical feasibility TBD)
- we publish updates to the Android app every week or so and request normal compensation requests each cycle on the DAO like normal, instead of doing the pre-payment thing

Of course we still have to define some milestones and figure out how this is going to work technically, but putting the details aside for now can we have your "concept ACK" to start working on the Android app with Bisq java GPRC as above? :sweat_smile:

**Comment by chimp1984** [comment]
@wiz 
Thanks for reconsidering that approach.
Here is my concept ACK ;-)

**Comment by chimp1984** [comment]
Here is a momory profile from the headless app running with JProfiler (250 MB):
<img width="1210" alt="Screen Shot 2019-11-13 at 18 18 30" src="https://user-images.githubusercontent.com/54558767/68812858-12b9dc00-0642-11ea-9ebe-88009825aaaf.png">

And that is with the UI app (350 MB):
<img width="1207" alt="Screen Shot 2019-11-13 at 18 20 23" src="https://user-images.githubusercontent.com/54558767/68812937-48f75b80-0642-11ea-916b-1f737045f891.png">

**Comment by chimp1984** [comment]
@wiz 
> I propose that we amend this DAO proposal as follows:

Could you include a link to that post in the initial first post so it is clear to readers and not get lost in the thread?

**Comment by wiz** [comment]
> Could you include a link to that post in the initial first post so it is clear to readers and not get lost in the thread?

Okay done.

As for the integration work, it was more straightforward to package the risq daemon inside the Android app because it was a native linux binary. Packing the Bisq Java app inside the Android app would be a little tricky because of the JVM, so instead I think we might try to run the headless Bisq entrypoint method inside a separate Android background service thread - obviously Java will compile for Android fine out of the box, we just gotta hack it to work with Android app lifecycle. After we figure that out we can begin the work of working to add the missing API methods, etc. and integrate the UI :grin:

**Comment by cbeams** [comment]
Exciting stuff, glad to see this effort coming together!

Regarding building the Android app on top of a new BIsq gRPC API, please take a look at the high-level plan for building that out at https://github.com/bisq-network/proposals/issues/136#issuecomment-554443705. Feedback welcome.

**Comment by cbeams** [comment]
> Packing the Bisq Java app inside the Android app would be a little tricky because of the JVM, so instead I think we might try to run the headless Bisq entrypoint method inside a separate Android background service thread

As a first cut, how about assuming that the user has a dedicated Bisq node running somewhere, and making RPC calls to it (possibly/probably over Tor)?

**Comment by wiz** [comment]
> As a first cut, how about assuming that the user has a dedicated Bisq node running somewhere, and making RPC calls to it (possibly/probably over Tor)?

We've been thinking about how to do the integration of Bisq Java codebase into the Android app. There are basically 2 ways to do it:

1) Run the headless Bisq daemon and connect over GRPC API using a localhost socket or over Tor to the desktop computer or Raspberry PI at home - this would allows either running Bisq locally or remotely.

2) Run the Bisq headless entrypoint inside an Android app's background service and post messages between the app thread and background service thread - this wouldn't work with remote nodes at home, but performance and UX of the Android app would likely be faster.

We need to do more research and testing into these architectures to figure out which one we will use, before we can answer the question...

**Comment by wiz** [comment]
Of course, maybe there is a hybrid solution where we can pass the GRPC messages between the background thread and the app thread, and have the option of sending the GRPC over a Tor socket to your computer at home... that would be the best of both worlds 😅

**Comment by objectorange** [comment]
One thing to be aware of with Android, not all Java will compile to Android's platform. I had this problem in 2018 with 1M5 as I'm using Neo4J and it would not work on Android. Definitely recommend a quick check to see if it can be compiled into an APK and installed on Android.

**Comment by wiz** [comment]
<img width="965" alt="Screen Shot 2020-06-13 at 20 17 47" src="https://user-images.githubusercontent.com/232186/84567355-161d6d80-adb3-11ea-8c55-2bed78464996.png">

This proposal was approved by DAO voting in Cycle 7. However, since the GRPC API is still not ready after 6 months of waiting, I will close this proposal as "stalled" for now. Hopefully we can still do the Android app someday if the GRPC API is completed.
